File: frontend/src/lib/accountability.ts
import type { DepartmentFigures, PublishingRecord, RecordPeriod, RecordRequirement } from "@/lib/api/types";

// The publishing record says what a gap is, and no more: a document not in the
// Ledger may exist and simply not be online, so a gap is "not found in the
// Documents Centre and not in The Ledger", never "does not exist".

export type RecordState = RecordPeriod["state"];

export const STATE_LABELS: Record<RecordState, string> = {
  held: "In The Ledger",
  related: "Related documents only",
  missing: "Not found",
  not_due: "Not yet expected",
};

const longDate = new Intl.DateTimeFormat("en-GB", { day: "numeric", month: "long", year: "numeric", timeZone: "UTC" });

export function dateLabel(isoDate: string): string {
  return longDate.format(new Date(`${isoDate.slice(0, 10)}T00:00:00Z`));
}

/** The one sentence a gap is allowed to say. */
export function gapSentence(checked: string): string {
  return `Not found in ama.gov.gh's Documents Centre (checked ${dateLabel(checked)}) and not in The Ledger.`;
}

export const ASSUMPTION_NOTE =
  "The “Expected…” dates under each document are our own conservative assumptions of when it is due, not statutory deadlines. The statutory dates under the Local Governance Act, 2016 (Act 936) should be confirmed before this record is used for real.";

export function yearsOf(record: PublishingRecord): number[] {
  return Array.from({ length: record.last_year - record.first_year + 1 }, (_, i) => record.first_year + i);
}

/** The periods a requirement shows under one year: one, four quarters, or the plan period covering it. */
export function periodsIn(requirement: RecordRequirement, year: number): RecordPeriod[] {
  if (requirement.cadence === "plan_period") {
    return requirement.periods.filter((p) => p.year <= year && year < p.year + 4);
  }
  return requirement.periods.filter((p) => p.year === year);
}

/** How many columns (years) a plan period spans in the table, from the first year shown. */
export function planSpan(period: RecordPeriod, years: number[]): number {
  return years.filter((year) => period.year <= year && year < period.year + 4).length;
}

/** The request someone can send under the RTI Act, worded for this document and period. */
export function requestWording(documentName: string, period: string, issuedElsewhere: boolean): string {
  const subject = issuedElsewhere ? `the ${documentName} on the Accra Metropolitan Assembly` : `the Accra Metropolitan Assembly's ${documentName}`;
  return `Under the Right to Information Act, 2019 (Act 989), I request a copy of ${subject} for ${period}.`;
}

export function rtiHref(requirement: RecordRequirement, period: RecordPeriod): string {
  const params = new URLSearchParams({ document: requirement.name, period: period.label });
  if (requirement.issued_by) params.set("elsewhere", "1");
  return `/rti?${params}`;
}

/** Whether a department had anything to show in the period: most never receive reports. */
export function isActive(department: DepartmentFigures): boolean {
  const counts = [department.reports.received, department.documents.accepted, department.documents.disputed, department.documents.auto_published];
  return counts.some((value) => value !== 0);
}

File: backend/app/schemas/accountability.py
"""Response models for the accountability pages: GET /api/publishing-record and GET /api/responsiveness."""

from typing import Literal

from pydantic import BaseModel

Count = int | None  # None: fewer than 5


class RecordDocument(BaseModel):
    id: str  # the Ledger document; its PDF is at /api/ledger/{id}/file
    title: str
    year: int | None
    year_source: Literal["confirmed", "cover", "title", "ledger"] | None  # ledger: from the Ledger's record, not the document
    department_name: str | None
    note: str | None  # why it counts as what it does, when its first page decided


class RecordPeriod(BaseModel):
    """One year, quarter or plan period of a required document."""

    label: str  # "2024", "Q2 2024", "2022–2025"
    year: int
    quarter: int | None
    # held: the document itself; related: documents about the same thing, not it; missing: neither, once
    # expected; not_due: not yet expected, by our assumption (not a statutory deadline)
    state: Literal["held", "related", "missing", "not_due"]
    expected_from: str | None  # the date it counts as expected, by our assumption
    documents: list[RecordDocument]
    related: list[RecordDocument]
    nearby: list[RecordDocument]  # what the Ledger holds that year from the same departments or categories
    nearby_total: int


class Requirement(BaseModel):
    id: str
    name: str
    cadence: Literal["annual", "quarterly", "plan_period", "as_issued"]
    issued_by: str | None  # when the Assembly isn't the one that issues it
    expected_note: str | None  # "Expected 6 months after the year ends": our assumption, not a statutory deadline
    nearby_scope: str  # where "what the Ledger holds" looks: departments and categories
    periods: list[RecordPeriod]  # empty for a document issued on no fixed schedule
    undated: list[RecordDocument]  # held, but its year isn't stated
    held: list[RecordDocument]  # for a document issued on no fixed schedule: every one held


class RequirementGroup(BaseModel):
    id: str
    name: str
    requirements: list[Requirement]


class RecordSummary(BaseModel):
    due: int
    held: int
    related: int
    missing: int
    not_due: int


class ReportingGap(BaseModel):
    """A figure the Assembly's documents once reported and haven't since: Nokware's reading, with its evidence."""

    id: str
    subject: str
    latest_year: int
    years_since: int
    figures: str  # the figures as the document gives them
    quote: str  # the passage they come from
    document_id: str  # its PDF is at /api/ledger/{id}/file
    document_title: str
    searched: list[str]  # the words the Ledger was searched for
    checked: str  # the date it was searched
    why: str


class PublishingRecord(BaseModel):
    generated_at: str
    documents_centre: str
    documents_centre_checked: str  # the date ama.gov.gh's Documents Centre was last checked
    first_year: int
    last_year: int
    groups: list[RequirementGroup]
    summary: RecordSummary
    gaps: list[ReportingGap]  # figures the documents once reported and haven't since
    gaps_about: str


class DepartmentReports(BaseModel):
    received: Count
    resolved: Count
    open: Count
    waiting: Count  # still not started this many days after it arrived (Responsiveness.waiting_days)
    median_days_to_start: float | None  # None until five
    median_days_to_resolve: float | None
    disputed: Count  # resolutions residents said weren't fixed
    confirmed: Count  # of those, the MCE confirmed the resolution
    reopened: Count  # or sent it back to be finished


class DepartmentDocuments(BaseModel):
    """Contributors' documents the department had 72 hours to review."""

    accepted: Count
    disputed: Count
    auto_published: Count  # the review clock ran out
    median_hours_to_review: float | None


class DepartmentFigures(BaseModel):
    id: str
    name: str
    reports: DepartmentReports
    documents: DepartmentDocuments


class MceFigures(BaseModel):
    documents_ruled: Count  # escalated disputes the MCE ruled on
    documents_run_out: Count  # or let the clock run out on
    reports_confirmed: Count
    reports_reopened: Count


class PetitionRefusals(BaseModel):
    reason: str
    label: str
    count: int


class PetitionFigures(BaseModel):
    """The MCE's handling of petitions: exact counts, since they count decisions on public petitions, not residents."""

    sent: int
    published_by_mce: int
    published_automatically: int  # the MCE let the 72 hours pass
    refused: int
    refusals: list[PetitionRefusals]
    reached_threshold: int  # the four below add up to this
    answered_in_time: int
    answered_late: int
    unanswered: int  # 30 days passed, no response yet
    waiting: int  # still within the 30 days


class Responsiveness(BaseModel):
    generated_at: str
    period_start: str
    waiting_days: int
    departments: list[DepartmentFigures]  # by name, never ranked; Assembly departments only
    mce: MceFigures
    petitions: PetitionFigures
